### PR TITLE
convert SDF to USD

### PR DESCRIPTION
**Pull request by adlarkin**
# 🎉 New feature

Related to #736 

## Summary
This targets the `sdf12` branch to add functionality for converting a SDF file to [USD](https://graphics.pixar.com/usd/release/index.html).

## Checklist
- [ ] Signed all commits for DCO
- [ ] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge**

### CODE REVIEW

File: src/usd/sdf_usd_parser/geometry.cc [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+    {
+      auto geomPrim = _stage->GetPrimAtPath(pxr::SdfPath(_path));
+      if (!geomPrim)
+      {
+        std::cerr << "Internal error: unable to get prim at path ["
+                  << _path << "], but a geom prim should exist at this path\n";
+        return false;
+      }
+
+      if (!pxr::UsdPhysicsCollisionAPI::Apply(geomPrim))
```
Comment (ahcorde) [inline src/usd/sdf_usd_parser/geometry.cc:286]: you should create two schemas, one for the visual and another one for the collision, even if this implies to duplicate the mesh.

Comment (adlarkin) [reply src/usd/sdf_usd_parser/geometry.cc:286]: I'm not sure that I understand what you mean - are you saying that the geometry should have two schemas applied to it (i.e., collision and something else)? What is the visual schema you mentioned, can you share a link to documentation/API somewhere?

**Comment by adlarkin** [comment]
As a note to myself and anyone else following the progress of this PR, here's what's supported as of commit https://github.com/ignitionrobotics/sdformat/pull/762/commits/64b5915447645b476c833037423ce68e323c068b:

* Models
    * static/non-static
    * links and joints (more info on these items below)
* Links
    * Mass (part of the link's `inertial`)
    * Visual
        * Geometry for `box`, `cylinder`, `sphere`, `capsule`, triangle meshes, and ground planes
    * Collisions for geometries defined above
* Joints
    * Basic support (`parent`, `child`, and `axis` information) for `revolute`, `ball`, `prismatic`, and `fixed` joint types

I still need to test the USD equivalents in simulation with Nvidia's isaac sim to make sure things behave as expected (so far, I have looked at the generated USD file via the [usd converter script](https://github.com/ignitionrobotics/sdformat/blob/adlarkin/sdf_to_usd/src/cmd/usdConverter.cc), and viewed converted objects with `usdview`). Here are a few SDF files that may be worth using for testing:
* [double pendulum](https://github.com/ignitionrobotics/ign-gazebo/blob/ign-gazebo6/examples/worlds/pendulum_links.sdf)
    * has links with basic geometry shapes and revolute joints
* [wind.sdf](https://github.com/ignitionrobotics/ign-gazebo/blob/ign-gazebo6/examples/worlds/wind.sdf)
    * has a joint with the parent being the world instead of a link (see `roofJoint`), and joints of type `fixed` and `ball`

File: src/cmd/usdConverter.cc [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+      std::cout << e << "\n";
+    return -2;
+  }
+
+  // only support SDF files with exactly 1 world for now
+  if (root.WorldCount() != 1u)
+  {
+    std::cerr << argv[1] << " does not have exactly 1 world\n";
+    return -3;
+  }
```
Comment (koonpeng) [inline src/cmd/usdConverter.cc:48]: Do you have an idea on how support for multi world sdformat look like?

I'm thinking one approach would be to create a root stage containing sub layers, a downside of this is that afaik the sub layers must be separate files, and that usd always composes the sublayers into a single scene.

Do you know how multi world sdformat is handled in ignition/gazebo? Or they composed into a single scene like usd, or is only one world is rendered at a time?

File: cmake/sdf_config.cmake.in [hunk 3]
```diff
@@ -37,6 +37,10 @@ find_package(ignition-math@IGN_MATH_VER@)
 list(APPEND @PKG_NAME@_INCLUDE_DIRS ${IGNITION-MATH_INCLUDE_DIRS})
 list(APPEND @PKG_NAME@_LIBRARY_DIRS ${IGNITION-MATH_LIBRARY_DIRS})
 
+find_package(ignition-commonh@IGN_COMMON_VER@)
```
Comment (koonpeng) [inline cmake/sdf_config.cmake.in:40]:
```suggestion
find_package(ignition-common@IGN_COMMON_VER@)
```
typo?

**Review by koonpeng — COMMENTED** [review]
I'm having problem building this with other ignition libraries, when building libraries that uses sdformat, I am getting linker errors

```
...
/usr/bin/ld: cannot find -lusdImagingGL
/usr/bin/ld: cannot find -lusdRiImaging
/usr/bin/ld: cannot find -lusdSkelImaging
/usr/bin/ld: cannot find -lusdVolImaging
/usr/bin/ld: cannot find -lusdAppUtils
collect2: error: ld returned 1 exit status
```

I think it's a combination of usd's cmake not being correct about the needed compiler flags, and that some libraries are not linking to sdformat via the target, but by `XXX_Libraries` variables, and that colcon is not correctly adding the workspace install directory to the list of default search directories.

Since #817 has been merged, updating this PR should fix the issue.

**Comment by adlarkin** [comment]
> I'm having problem building this with other ignition libraries

@koonpeng, sorry, we actually are not iterating on this PR anymore - now that we have merged #817, the approach being taken is to split this PR up into smaller pieces that can be built on top of #817. For example:
* #829 
* #830 

We are keeping this PR open as a draft for reference, but once we have broken this up into smaller pieces that make use of the component infrastructure, we will close this PR.

**Comment by koonpeng** [comment]
thanks, I will look at #829 for my testing.

**Comment by ahcorde** [comment]
This PR was superseded by:
 - https://github.com/ignitionrobotics/sdformat/pull/828
 - https://github.com/ignitionrobotics/sdformat/pull/829
 - https://github.com/ignitionrobotics/sdformat/pull/830
 - https://github.com/ignitionrobotics/sdformat/pull/831
 - https://github.com/ignitionrobotics/sdformat/pull/836
 - https://github.com/ignitionrobotics/sdformat/pull/837
 - https://github.com/ignitionrobotics/sdformat/pull/849
 - https://github.com/ignitionrobotics/sdformat/pull/850

Thank you @scpeters @adlarkin @azeey @koonpeng for the reviews!

**Pull request closed without merging**